File: packages/core-ts/src/contract.ts
/**
 * The contract between a hand-written core and the code `truewire generate typescript`
 * emits: the TypeScript half of `truewire_core.contract`.
 *
 * A generated endpoint class takes its core as a constructor argument and calls exactly
 * one verb on it; a generated router hands the same core to every child. The generator
 * reads nothing else from a core and never imports the project's own `core/` module: the
 * interfaces below are structural, so a core satisfies them by shape and `tsc` checks the
 * two against each other where the client is constructed.
 *
 * `Meta` is the per-endpoint `meta` shape the core's `[cores.<name>]` entry declares in
 * `truewire.toml`, rendered by the generator into `<package>/meta.ts`; a core with no
 * schema receives `{}`.
 *
 * Verbs:
 *
 * - `request({ method, path, ... })`: one HTTP call (`HttpEndpoint`).
 * - `request({ path, ... })`: one WebSocket command/reply call, `path` being the wire
 *   method name (`CommandEndpoint`).
 * - `subscribe({ channel, parameters, ... })`: one channel subscription, returning a
 *   `Subscription` the caller awaits or iterates (`StreamEndpoint`).
 */
import type { Codec } from './validation.js'
import type { Subscription } from './ws/streams.js'

/** Options every generated method takes as its last parameter. */
export interface CallOptions {
  /** Override this call's response validation; the client-level default when omitted. */
  validate?: boolean
  /** Abort the call. */
  signal?: AbortSignal
}

/** What every verb carries: the generated request value, its codec, the reply's codec and the endpoint's `meta`. */
export interface Call<Req, Res, Meta> extends CallOptions {
  /**
   * The generated request value: an object whose keys are the wire's own parameter
   * names (a `Request` interface), a union member, an array, or `undefined` when the
   * endpoint declares no request.
   */
  request: Req | undefined
  /** Codec of `request`, for `dump` to the wire; `undefined` with no request. */
  requestCodec: Codec<Req> | undefined
  /** Codec of the value the method returns, for `parse`; `undefined` when it returns nothing. */
  responseCodec: Codec<Res> | undefined
  /** The endpoint's declared `meta`, in the shape its core's schema states. */
  meta: Meta
}

/** One HTTP call. `{name}` placeholders in `path` are filled from `request`. */
export interface HttpCall<Req, Res, Meta> extends Call<Req, Res, Meta> {
  /** The wire HTTP method; `undefined` when the spec leaves it to the core (a uniformly POST JSON-RPC API). */
  method: string | undefined
  /** The wire path template. */
  path: string
}

/** One WebSocket command; `path` is the wire method name. */
export interface CommandCall<Req, Res, Meta> extends Call<Req, Res, Meta> {
  path: string
}

/** One channel subscription; `{name}` placeholders in `channel` are filled from `parameters`. */
export interface SubscribeCall<Params, Message, Meta> extends CallOptions {
  channel: string
  parameters: Params | undefined
  parametersCodec: Codec<Params> | undefined
  /** Codec of each pushed message. */
  messageCodec: Codec<Message> | undefined
  meta: Meta
}

/**
 * Base of a generated `rpc` endpoint reached over HTTP.
 *
 * `request` sends one call and returns the value the method's return type describes:
 * `responseCodec.parse` of the reply's JSON when validation is on, the raw `JSON.parse`
 * value (cast) when `validate` is `false`.
 */
export interface HttpEndpoint<Meta = Record<string, never>> {
  request<Req, Res>(call: HttpCall<Req, Res, Meta>): Promise<Res>
}

/** Base of a generated `rpc` endpoint reached over a WebSocket connection. */
export interface CommandEndpoint<Meta = Record<string, never>> {
  request<Req, Res>(call: CommandCall<Req, Res, Meta>): Promise<Res>
}

/** Base of a generated `stream` endpoint. */
export interface StreamEndpoint<Meta = Record<string, never>> {
  subscribe<Params, Message>(call: SubscribeCall<Params, Message, Meta>): Subscription<Message>
}

File: crates/truewire-core/src/contract.rs
//! The contract between a hand-written core and the code `truewire generate rust` will
//! emit: the Rust half of `truewire_core.contract` and `@truewire/core`'s `contract.ts`.
//!
//! A generated endpoint struct holds its core as an `Arc<dyn HttpEndpoint<Meta>>` (or
//! `CommandEndpoint`/`StreamEndpoint`) and calls exactly one verb on it; a generated router
//! hands the same core to every child. The generator reads nothing else from a core and
//! never imports the project's own `core` module: the traits below are what a core
//! implements, and the compiler checks the two against each other where the client is
//! constructed (ADR 0011).
//!
//! `Meta` is the per-endpoint `meta` shape the core's `[cores.<name>]` entry declares in
//! `truewire.toml`, rendered by the generator into `<package>/meta.rs`; a core with no
//! schema receives `()`.
//!
//! What the plan carries, and where it goes:
//!
//! - `wire.method`/`wire.path` (an HTTP endpoint) or `wire.path` (a WebSocket command's
//!   method name) or `wire.channel` (a stream): on the call, verbatim, placeholders unfilled.
//! - The request: dumped through `serde` into a wire [`Value`] before the call, so the
//!   core sees wire names and wire forms (a `TimestampMillis` is already an integer) and
//!   fills `{name}` placeholders from it. `None` when the endpoint declares no request.
//! - The reply: the core returns the wire body as a [`Value`], after unwrapping its
//!   envelope (`envelope.payload`) and mapping errors. Generated code then either
//!   [`decode`](crate::validation::decode)s it into the response type or, for
//!   `validate: false`, hands the `Value` back as it came. That is why the traits are not
//!   generic in the response type and why a core never validates: `validate` is a decision
//!   the generated method takes after the call, not a flag the core reads.
//!
//! Verbs:
//!
//! - [`HttpEndpoint::request`]: one HTTP call.
//! - [`CommandEndpoint::request`]: one WebSocket command/reply call.
//! - [`StreamEndpoint::subscribe`]: one channel subscription, returning a
//!   [`Stream`] of pushed payloads the caller iterates and unsubscribes.
//!
//! There is no `ClientRoot` or `Composite` trait: a Rust root is a struct the generator
//! writes with one field per transport the `truewire.toml` core declares, and the
//! hand-written code builds it (`Client::new(core)`); nothing needs a `new(...)` protocol.

use std::time::Duration;

use async_trait::async_trait;
use serde_json::Value;

use crate::errors::Result;
use crate::ws::Stream;

/// Options every generated method takes as its last parameter.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct CallOptions {
    /// Time before this call is abandoned with a `NetworkError`; the core's default when `None`.
    pub timeout: Option<Duration>,
}

impl CallOptions {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn timeout(mut self, timeout: Duration) -> Self {
        self.timeout = Some(timeout);
        self
    }
}

/// One HTTP call. `{name}` placeholders in `path` are filled from `request`.
#[derive(Debug, Clone)]
pub struct HttpCall<'a, Meta> {
    /// The wire HTTP method; `None` when the spec leaves it to the core (a uniformly POST JSON-RPC API).
    pub method: Option<&'a str>,
    /// The wire path template.
    pub path: &'a str,
    /// The dumped request: an object keyed by the wire's own parameter names, a union
    /// member, an array, or `None` when the endpoint declares no request.
    pub request: Option<Value>,
    /// The endpoint's declared `meta`, in the shape its core's schema states.
    pub meta: &'a Meta,
    pub options: CallOptions,
}

/// One WebSocket command; `path` is the wire method name.
#[derive(Debug, Clone)]
pub struct CommandCall<'a, Meta> {
    pub path: &'a str,
    pub request: Option<Value>,
    pub meta: &'a Meta,
    pub options: CallOptions,
}

/// One channel subscription; `{name}` placeholders in `channel` are filled from `parameters`.
#[derive(Debug, Clone)]
pub struct SubscribeCall<'a, Meta> {
    pub channel: &'a str,
    /// The dumped parameters object, or `None` for a direct-channel or connect-only
    /// stream, where the generated method filled the channel itself.
    pub parameters: Option<Value>,
    pub meta: &'a Meta,
    pub options: CallOptions,
}

/// Base of a generated `rpc` endpoint reached over HTTP.
#[async_trait]
pub trait HttpEndpoint<Meta = ()>: Send + Sync {
    /// Send one call and return the wire body the method's return type describes, its
    /// envelope unwrapped and its errors mapped.
    async fn request(&self, call: HttpCall<'_, Meta>) -> Result<Value>;
}

/// Base of a generated `rpc` endpoint reached over a WebSocket connection.
#[async_trait]
pub trait CommandEndpoint<Meta = ()>: Send + Sync {
    /// Send one command and return its reply.
    async fn request(&self, call: CommandCall<'_, Meta>) -> Result<Value>;
}

/// Base of a generated `stream` endpoint.
#[async_trait]
pub trait StreamEndpoint<Meta = ()>: Send + Sync {
    /// Subscribe to `channel` with `parameters`; each pushed payload is one item.
    async fn subscribe(&self, call: SubscribeCall<'_, Meta>) -> Result<Stream<Value>>;
}
